src/main/kotlin/org/dashevo/dapiclient/DapiClient.kt

```diff
@@ -1,219 +1,398 @@
 import retrofit2.converter.gson.GsonConverterFactory
-import java.util.concurrent.TimeUnit
+import java.util.logging.Level
+
+
+/**
+ * A simple client that requests a greeting from the [HelloWorldServer].
+ */
+class DapiClient(val masternodeService: MasternodeService, val diffMasternodeEachCall: Boolean = true, val debug: Boolean = false) {
+
+    // gRPC properties
+    lateinit var channel: ManagedChannel
+    lateinit var platform: PlatformGrpc.PlatformBlockingStub
+    lateinit var core: CoreGrpc.CoreBlockingStub
+    private var initialized = false
+
+    // jRPC Properties
+    private lateinit var retrofit: Retrofit
+    private lateinit var dapiService: DapiService
+    private val debugOkHttpClient: OkHttpClient
+    private var initializedJRPC = false
 
-open class DapiClient(mnIP: String, mnDapiPort: String, debug: Boolean = false) {
+    // Constants
+    companion object {
+        private val logger = Logger.getLogger(DapiClient::class.java.name)
+        const val DEFAULT_GRPC_PORT = 3010
+        const val DEFAULT_JRPC_PORT = 3000
+    }
 
-    private val retrofit: Retrofit
 
-    val dapiService: DapiService
 
     init {
         val logging = HttpLoggingInterceptor()
         logging.level = HttpLoggingInterceptor.Level.BODY
 
-        val debugOkHttpClient = OkHttpClient.Builder()
+        debugOkHttpClient = OkHttpClient.Builder()
                 .addInterceptor(logging)
-                .connectTimeout(1, TimeUnit.MINUTES)
-                .readTimeout(1, TimeUnit.MINUTES)
-                .writeTimeout(1, TimeUnit.MINUTES)
-                .build()
-        retrofit = Retrofit.Builder()
-                .addConverterFactory(GsonConverterFactory.create())
-                .baseUrl("$mnIP:$mnDapiPort/")
-                .client(if (debug) debugOkHttpClient else OkHttpClient())
+                .connectTimeout(10, TimeUnit.SECONDS)
+                .readTimeout(10, TimeUnit.SECONDS)
+                .writeTimeout(10, TimeUnit.SECONDS)
                 .build()
-        dapiService = retrofit.create(DapiService::class.java)
+
     }
 
-    inline fun <reified T> Gson.fromJson(json: JsonElement): T = this.fromJson<T>(json, object : TypeToken<T>() {}.type)
+    constructor(masternodeAddress: String, diffMasternodeEachCall: Boolean = true):
+            this(SingleMasternode(masternodeAddress), diffMasternodeEachCall)
+
+    /* Platform gRPC methods */
 
     /**
-     * Create a DAP
-     * @param dapSchema Schema for the DAP being Created
-     * @param userRegTxId hash of Blockchain User registration transaction (SubTx)
-     * @param cb callback, instance of {@link DapCallback}
+     * Send State Transition to machine
+     *
+     * @param stateTransition
      */
-    open fun registerDap(dapSchema: Map<String, Any>, userRegTxId: Sha256Hash, hashPrevSubTx: Sha256Hash,
-                         privKey: ECKey, cb: DapiRequestCallback<String>) {
+    fun applyStateTransition(stateTransition: StateTransition) {
```

Comment:
This method has no return value or callback parameter, how is it intended to be used?

Reply:
An exception is thrown if there is a failure (not connected, invalid arguments).  Right now its up to the calling method to handle this.  I was thinking about catching the exception, then parsing it and throwing a specific error exception that is part of DPP.

The "DashPay" library will be calling this method and I haven't worked out all the details yet on how it will handle such errors.  I will check the DashJS and DashSync libraries to see how they do it.

Reply:
The method according to the javascript does return an empty object

```javascript
/**
   * Send State Transition to machine
   *
   * @param {DataContractStateTransition|DocumentsStateTransition} stateTransition
   * @returns {Promise<!ApplyStateTransitionResponse>}
   */
  async applyStateTransition(stateTransition) {
```

where 

```protobuf
message ApplyStateTransitionResponse {

}
```

With other methods the values returned are the relevant data and in this case the return value is the same as a `void` in java.  

The only way to truly know of success is to query the platform based on the type using `getIdentity`, `getDocuments`, `getContract` and in DashSync within the BlockchainIdentity class that is how it is handled.  The client must keep asking to if the identity, contract or documents exist (if the transition was to CREATE).  

In JavaScript all of these methods are written as asynchronous methods, but they are commonly called with `await`.  Thus far, all of the kotlin methods are written as synchronous or blocking methods.  However, there is support within the grpc library to make calls that are asynchronous (non-blocking).  I figured that in the App we would have a background task that is executed as part of the ViewModel and call an observer when the background task is complete with success or fails with an error.